README.md
# node_smartystreets

A high-performance client for the SmartyStreets geocoding api

## Dependencies

node and npm

## Installation

Depending on your environment you may need to run this as root in order to get the command line tool

`npm install -g smartystreets`

## Command line tool

Usage: smartystreets [options]

```
Options:

  -h, --help                      output usage information
  -V, --version                   output the version number
  -i, --input [file]              Input (csv file) [stdin]
  -o, --output [file]             Output (csv file) [stdout]
  -s, --street-col [col_csv]      Street col CSV [street]
  -z, --zipcode-col [col]         Zipcode col [zipcode]
  -c, --city-col [col]            City col [city]
  -S, --state-col [col]           State col [state]
  --zipcode-filter [zips]         Only geocode records in certain zipcodes, comma separated
  --state-filter [states]         Only geocode records in certain states, comma separated
  -d, --delimiter [symbol]        CSV delimiter in input file
  -O, --output-split [column]     Write to multiple files divided by column
  --truncate-split [length]       Used with --output-split, truncate column to first X characters
  -a, --auth-id [id]              SmartyStreets auth id [environment variable smartystreets_auth_id]
  -A, --auth-token [token]        SmartyStreets auth token [environment variable smartystreets_auth_token]
  -j, --concurrency [jobs]        Maximum number of concurrent requests [48]
  -✈, --column-definition [mode]  Column definition mode or file [standard]
  -p, --column-prefix [text]      Prefix for smartystreets columns in the output file [ss_]
  -x, --column-suffix [text]      Suffix for smartystreets columns in the output file
  -q, --quiet                     Quiet mode - turn off progress messages
  -l, --log-interval [num]        Show progress after every X number of rows [1000]
  --drop-threshold [rows]         Maximum number of rows that can be dropped due to api failures [Infinity]
  --include-invalid               Allows approximate geocoding of records that SmartyStreets considers invalid
```

### parameters

`-i, --input [file]`

Not required. Default stdin. If included it should be a path to a csv file.

`-o, --output [file]`

Not required. Default stdout. If included it should be a path to a csv file that does not exist yet.

`-s, --street-col [col_csv]`

Not required. Default is 'street'. It should be the name of the column in your input file that contains the first part of the address, e.g. '123 main st'. You can pass a CSV like "HOUSE,PREDIR,STREET,STRTYPE,POSTDIR,APTTYPE,APTNBR" to auto-combine fields (spaces will be added). Note that, currently, spaces or special characters are not allowed in the CSV.

`-z, --zipcode-col [col]`

Not required. Default is 'zipcode'. It should be the name of the column in your input file that contains the zipcode of an address, e.g. '12345'

_Note_: if your file includes street and zipcode, you don't need to include city and state data, as it will not be used. If your file does not include zipcode, you _must_ include city _and_ state columns.

`-c, --city-col [col]`

Not required. Default is 'city'. It should be the name of the column in your input file that contains the city or town, e.g. 'Albany'

_Note_: city is only used if zipcode is not present

`-S, --state-col [col]`

Not required. Default is 'state'. It should be the name of the column in your input file that contains the state, e.g. 'Vermont'

`--zipcode-filter [zips]`

Not required. Only geocodes records in certain zipcodes. Separate by comma for multiple zips. Example: `--zipcode-filter 05401,05403`. Uses column specified in `--zipcode-col`.

`--state-filter [zips]`

Not required. Only geocodes records in certain states. Separate by comma for multiple states. Example: `--state-filter VT,NY`. Uses column specified in `--state-col`.

`-O, --output-split [column]`

Not required. This will write multiple output files, split by a column value. `--output-split ss_zipcode` will write one file per zipcode, under `output/12345.csv`.

`--truncate-split [length]`

Not required. Used with `--output-split`. It will truncate the column value to a number of characters. `--output-split ss_zipcode --truncate-split 3` will write files like `output/123.csv` which will contain all rows with zipcodes that start with 123.

`-a, --auth-id [id]`

_Required_. You can get this [here](https://smartystreets.com/account/keys) listed under "Security Keys"

`-A, --auth-token [token]`

_Required_. You can get this [here](https://smartystreets.com/account/keys) listed under "Security Keys"

`-j, --concurrency [jobs]`

Not required. Default 48. Tune this if you fancy yourself an engineer and want to get a slight increase in performance. It should be higher if you have an unmetered 10 gigabit connection, and lower if you're on an EDGE wireless signal. If you want to run fifty instances of this program at the same time, you may experience dropped connections which can be dealt with by turning this down.

`-✈, --column-definition [mode|file|string]`

Not required. Default standard. This defines which columns from smartystreets will be added to your output file. You can find the definitions in the `structure` directory. `mail` will get you a mailing address and nothing more, and `complete` will get you every column which smartystreets has to offer. You can also make your own custom definition based on one of the existing definitions and use that by passing in the path. Also, if you're insane, you can just pass it JSON directly. I don't even know why I added that option. Probably had a good reason at the time. Don't even know if it works. I should add tests.

`-p, --column-prefix [text]`

Not required. Default ss_. This is the prefix applied to all columns that are added to your output file. `delivery_line_1` becomes `ss_delivery_line_1`

`-x, --column-suffix [text]`

Not required. Default empty string. This is the suffix applied to all columns that are added to your output file. `delivery_line_1` could become `delivery_line_1_suffixGoesHere`

`-q, --quiet`

For when you just want it to shut up already

`-l, --log-interval [num]`

This will log a message every X number of rows to let you know how it's doing. The default is 1,000 and you can configure it with this option.

`--drop-threshold [rows]`

If a batch of rows cannot be geocoded after 5 retries, they will be dropped. You can use this option to limit that. If you set `--drop-threshold 0`, any dropped rows will cause the process to exit with a code of 1. If you set it to 1,000, geocoding failures will be tolerated up to 1,000 rows, then the process will exit with a code of 1.

`--include-invalid`

Not required. Allows approximate geocoding of addresses that SmartyStreets considers invalid. Be cautious - this may introduce potentially-large precision degredation.

## Geocoding stream API

This module also provides a simple api that you can use in your own node projects.

```
var Smartystreets = require('smartystreets');
var fs = require('fs');
var csv = require('fast-csv');

var geocoder = new Smartystreets(options);

fs.createReadStream('input.csv')
  .pipe(csv({ headers: true }))
  .pipe(Smartystreets.grouper(70))
  .pipe(geocoder)
  .pipe(process.stdout);
```

In this example, `geocoder` is a duplex stream which will take every address that is written to it in CSV format and output a geocoded CSV.

The `options` argument is identical to the options used by the command line tool, except that camelCase is used (street-col -> streetCol) and "input" and "output" are ignored (instead use the piping solution above).

For more API details, see [our TypeScript API declarations][API].

[API]: https://github.com/faradayio/node_smartystreets/blob/master/index.d.ts

## Notes

* This is designed for performance and if you run it on the wrong file you might use up your entire api quota in a matter of seconds. If you have a limited number of lookups, be careful.
* Since many requests are made in parallel, it's unlikely that the rows of the output CSV will be in the same order as the input CSV.

## Authors

* [Tristan Davies](mailto:npm@tristan.io)
* [Seamus Abshere](mailto:seamus@abshere.net)

## Corporate support

<p><a href="http://faraday.io" alt="Faraday"><img src="https://s3.amazonaws.com/creative.faraday.io/logo.png" alt="Faraday logo"/></a></p>

## Copyright

Copyright 2015 [Faraday](http://faraday.io)


CHANGELOG.md
## 3.0.2 (2019-08-07)

* Bug fixes

  * Fix "freeform" input mode

## 3.0.1 (2019-02-05) - BROKEN

* Bug fixes

  * Always include everything, even if --include-invalid isn't passed

## 3.0.0 (2019-02-05) - BROKEN

[nothing]

## 2.0.2 (2019-02-05) - BROKEN

* Bug fixes

  * Fix dups due to missed callback

## 2.0.1 (2018-11-25) - BROKEN

* Bug fixes

  * Fix dups due to bad fast-csv

## 2.0.0 (2018-11-19) - BROKEN

* Breaking changes

  * Include all input rows in output

## 1.8.2 (2017-07-28)

* Re-publishing

## 1.8.1 (2017-07-28) BROKEN

* Bug fixes

  * Don't print progress to standard out
  * Be a bit more aggressive about CLI argument validation

* Enhancements

  * Convert code to TypeScript & ES6 syntax internally

## 1.8.0 (2017-07-22)

* Bug fixes

  * Fixed columnList concurrency issue when running as a lib

* Enhancements

  * Option outputStreamFormat: "object"
  * Export grouper (used by CLI, now avail for use in libs)

## 1.7.9 (2017-07-21)

Add TypeScript declarations.

## 1.7.8 (2017-07-19)

Fix bug in --street-col

## 1.7.7 (2017-07-19)

Allow --street-col "HOUSE,PREDIR,STREET,STRTYPE,POSTDIR,APTTYPE,APTNBR".

## 1.7.6 (2017-05-15)

Fix loss of rows when splitting.

## 1.7.2 (2015-12-28)

Remove the workaround added in 1.5.1 since that issue has been fixed

## 1.7.0 (2015-8-14)

Add --drop-threshold, a solution for issue #31

## 1.6.2 (2015-7-16)

fix issue #27, allowing geocoding to happen from a single address column

## 1.6.1 (2015-7-14)

fix bug where using --output-split on a column that had nonsafe characters in it would break

## 1.6.0 (2015-7-14)

add --retry-timeout option

## 1.5.1 (2015-6-11)

workaround for memory leak in request module (https://github.com/request/request/issues/1637)

## 1.5.0 (2015-5-26)

add --include-invalid option

## 1.4.1 (2015-5-26)

update dependencies, filter rows for validity before attempting to geocode

## 1.3.2 (2015-4-27)

Switch to faradayio/forever-agent fork to prevent weird node 0.12 bug

## 1.3.1 (2015-4-27)

Workaround for https://github.com/npm/npm/issues/7773

## 1.3.0 (2015-4-26)

Fixed request forever/pool issue in node 0.12 https://github.com/request/request/issues/1511

Updated dependencies

## 1.2.0 (2015-4-20)

Features:

  - --truncate-split [len] option. When used with --output-split, it will split by the first X characters in a cell. Example: `--output-split zipcode --truncate-split 3` will split by the first 3 digits of the zipcode column.
  - --state-filter [state] option. Use it to only geocode records in a certain state. Eg `--state-filter VT`. Uses your state column as specified in `--state-col`.
  - --zipcode-filter [zipcode] option. Same as above, but with zipcodes.

Bugs:

  - `--output-split` no longer broken

## 1.0.1 - 1.1.0 (2015-3-05 - 2015-4-20)

Fixes a major memory leak in --output-split but due to a mistake in which code was commited to the wrong branch, also breaks output-split. Avoid these versions, use 1.3.0 instead. Oops.

## 1.0.0 (2015-2-27)

Fix issue #30, some extra stability

## 1.0.0-rc.* (a long time)

Playing whack-a-bug

## 1.0.0-rc.0 (2015-01-02)

Major refactor using through2 streams for cleaner and safer code

Features:

  - `--output-split [column]` option. Writes to multiple output files based on a column name. Can be a column that is fetched from smartystreets, such as state. If specified, `--output` will be a directory instead of a file.
  - `--log-interval [rows]` option. Log every X number of rows. Was previously hard-coded to 1000. Default is 1000.
  - `--quiet` option. Actually works now. Suppresses non-fatal errors and progress messages.

Bugs:

  - Elusive memory bugs sorted out by switching to through2 and thus preventing backpressure issues caused by poorly implemented streams

Features removed:

  - Redis caching. Who needs it? Not me. It was a bad idea anyways.

## 0.8.0 (2014-12-30)

Features:

 - `--delimiter` option

## 0.7.0 (2014-09-12)

Features:

 - Works with all-in-one street columns like "123 n blount st, madison, wi 53703" (issue #26)

## Previous

See [github commits](https://github.com/faradayio/node_smartystreets/commits/master/package.json)


program.ts
#!/usr/bin/env node

import csv = require('fast-csv');
import fs = require('fs');
import options = require('commander');
import path = require('path');
import through2 = require('through2');
import sanitizeFilename = require('sanitize-filename');

import grouper from './src/grouper';
import geocoder from './src/geocoder';
import { DEFAULTS, FullOptions, OutputStreamFormat } from './src/options';
import streamSplitter from './src/streamSplitter';

const packageJson = JSON.parse(fs.readFileSync(path.join(__dirname, '../package.json'), 'utf8'));

options
  .version(packageJson.version)
  .option('-i, --input [file]', 'Input (csv file) [stdin]', '-')
  .option('-o, --output [file]', 'Output (csv file) [stdout]', '-')
  .option('-s, --street-col [col_csv]', 'Street col CSV [street]', DEFAULTS.streetCol)
  .option('-z, --zipcode-col [col]', 'Zipcode col [zipcode]', DEFAULTS.zipcodeCol)
  .option('-c, --city-col [col]', 'City col [city]', DEFAULTS.cityCol)
  .option('-S, --state-col [col]', 'State col [state]', DEFAULTS.stateCol) // short forms are dumb
  .option('--zipcode-filter [zips]', 'Only geocode records in certain zipcodes, comma separated', '')
  .option('--state-filter [states]', 'Only geocode records in certain states, comma separated', '')
  .option('-d, --delimiter [symbol]', 'CSV delimiter in input file', ',')
  .option('-O, --output-split [column]', 'Write to multiple files divided by column', '')
  .option('--truncate-split [length]', 'Used with --output-split, truncate column to first X characters', 0)
  .option('-a, --auth-id [id]', 'SmartyStreets auth id [environment variable smartystreets_auth_id]', process.env.SMARTYSTREETS_AUTH_ID)
  .option('-A, --auth-token [token]', 'SmartyStreets auth token [environment variable smartystreets_auth_token]', process.env.SMARTYSTREETS_AUTH_TOKEN)
  .option('-j, --concurrency [jobs]', 'Maximum number of concurrent requests [48]', DEFAULTS.concurrency)
  .option('-✈, --column-definition [mode]', 'Column definition mode or file [standard]', 'standard')
  .option('-p, --column-prefix [text]', 'Prefix for smartystreets columns in the output file [ss_]', DEFAULTS.columnPrefix)
  .option('-x, --column-suffix [text]', 'Suffix for smartystreets columns in the output file', DEFAULTS.columnSuffix)
  .option('--include-invalid', 'Activates agressive matching by setting the header X-Include-Invalid to true')
  .option('-q, --quiet', 'Quiet mode - turn off progress messages')
  .option('-l, --log-interval [num]', 'Show progress after every X number of rows [1000]', DEFAULTS.logInterval)
  .option('--retry-timeout [num]', 'Retry failed requests after X milliseconds [30000]', DEFAULTS.retryTimeout)
  .option('--drop-threshold [rows]', 'Maximum number of rows that can be dropped due to api failures [Infinity]', 'infinity')
  .parse(process.argv);

options.concurrency = parseInt(options.concurrency);
options.truncateSplit = parseInt(options.truncateSplit);

if (!options.concurrency) {
  console.error('Invalid concurrency');
  options.help();
}

if (!options.authId) {
  console.error('Please specify a SmartyStreets auth id');
  options.help();
}

if (!options.authToken) {
  console.error('Please specify a SmartyStreets auth token');
  options.help();
}

if (!options.input) {
  console.error('Please specify an input file');
  options.help();
} else {
  if (options.input == '-' && !options.quiet) {
    console.error('***********************************************')
    console.error('Processing from STDIN')
    console.error('***********************************************')
  } else if (!options.quiet) {
    console.error('***********************************************')
    console.error('Processing ' + options.input)
    console.error('***********************************************')
  }
}
if (!options.output) {
  console.error('Please specify an output file');
  options.help();
}

if (typeof options.retryTimeout === 'string') {
  options.retryTimeout = Number(options.retryTimeout)
}

if (typeof options.dropThreshold !== 'string' || options.dropThreshold.toLowerCase() === 'infinity') {
  options.dropThreshold = null;
} else {
  options.dropThreshold = Number(options.dropThreshold);
  if (options.dropThreshold < 0 || options.dropThreshold % 1 !== 0) {
    console.error('invalid drop threshold, if specified it must be a positive integer or zero');
    options.help();
  }
}

options.streetCol = (options.streetCol && options.streetCol.indexOf(',') !== -1) ? options.streetCol.split(',') : options.streetCol;
const filterOptions = {
  zipcodeFilter: options.zipcodeFilter.length ? options.zipcodeFilter.split(',') : false,
  stateFilter: options.stateFilter.length ? options.stateFilter.split(',') : false,
}

const columnModes = ['mail', 'standard', 'complete', 'basic'];
if (columnModes.indexOf(options.columnDefinition) != -1) {
  //modes are in structure/mode.json
  options.structure = JSON.parse( fs.readFileSync(__dirname+'/../structure/'+options.columnDefinition+'.json', 'utf8') );
} else if (options.columnDefinition.substr(0, 1) == '{') {
  //it's a JSON object
  options.structure = JSON.parse(options.columnDefinition);
} else {
  //it's a JSON file
  options.structure = JSON.parse( fs.readFileSync(options.columnDefinition, 'utf8') );
}

// Default booleans to `false` instead of `undefined`.
if (!options.quiet) {
  options.quiet = false
}
if (!options.includeInvalid) {
  options.includeInvalid = false
}

const readStream = options.input == '-' ? process.stdin : fs.createReadStream(options.input);

let writeStream;
if (options.outputSplit) {
  try {
    fs.mkdirSync(options.output);
  } catch(e) {
    if ( e.code != 'EEXIST' ) throw e;
  }

  let firstRow = true, rowIndex: number, headers: string | undefined;

  writeStream = streamSplitter(function(row){
    if (firstRow) {
      firstRow = false;
      rowIndex = row.indexOf(options.outputSplit);
      headers = row;

      if (rowIndex == -1) {
        console.error('couldn\'t find column', options.outputSplit, 'in', row);
        process.exit();
      }

      return false;
    }
    let cell = String(row[rowIndex] || '');

    if (options.truncateSplit !== 0) {
      cell = cell.substr(0, options.truncateSplit);
    }

    return cell;
  }, function(streamName){
    const source = csv.createWriteStream();
    if (typeof headers !== 'undefined') {
      source.write(headers);
    } else {
      throw new Error("Could not find headers in stream")
    }
    source.pipe( fs.createWriteStream(options.output+'/'+sanitizeFilename(streamName)+'.csv') );
    return source;
  }, {objectMode: true});
} else if (options.output == '-') {
  writeStream = csv.createWriteStream();
  writeStream.pipe( process.stdout );
} else {
  writeStream = csv.createWriteStream();
  writeStream.pipe( fs.createWriteStream(options.output) );
}

/**
 * Verify that an option has the requested type. The caller is responsible for
 * setting `T` and `type` in a compatible fashion.
 */
function checkOpt<T>(options: { [key: string]: any }, key: string, type: string | string[]): T {
  if (typeof options[key] === type)
    return options[key]
  else
    throw new Error(`Expected ${key} to be of type ${type}, got ${JSON.stringify(options[key])}`)
}

function checkStreetCol(streetCol: any): string | string[] {
  if (typeof streetCol === 'string' || Array.isArray(streetCol))
    return streetCol
  else
    throw new Error(`Expected streetCol to be of type string or string[], got ${JSON.stringify(streetCol)}`)

}

function checkDropThreshold(dropThreshold: any): number | null {
  if (typeof dropThreshold === 'number' || dropThreshold === null)
    return dropThreshold
  else
    throw new Error(`Expected dropThreshold to be of type string or null, got ${JSON.stringify(dropThreshold)}`)
}

// Make sure we pass all `options` fields, and that all values have the correct
// type. This is mostly to ensure that our CLI and our API actually match, but
// it may also catch some invalid CLI options, too.
const geocodeOptions: FullOptions = {
  authId: checkOpt<string>(options, 'authId', 'string'),
  authToken: checkOpt<string>(options, 'authToken', 'string'),
  structure: checkOpt<object>(options, 'structure', 'object'),
  streetCol: checkStreetCol(options.streetCol),
  zipcodeCol: checkOpt<string>(options, 'zipcodeCol', 'string'),
  cityCol: checkOpt<string>(options, 'cityCol', 'string'),
  stateCol: checkOpt<string>(options, 'stateCol', 'string'),
  concurrency: checkOpt<number>(options, 'concurrency', 'number'),
  columnPrefix: checkOpt<string>(options, 'columnPrefix', 'string'),
  columnSuffix: checkOpt<string>(options, 'columnSuffix', 'string'),
  quiet: checkOpt<boolean>(options, 'quiet', 'boolean'),
  logInterval: checkOpt<number>(options, 'logInterval', 'number'),
  dropThreshold: checkDropThreshold(options.dropThreshold),
  retryTimeout: checkOpt<number>(options, 'retryTimeout', 'number'),
  includeInvalid: checkOpt<boolean>(options, 'includeInvalid', 'boolean'),
  outputStreamFormat: "array",
}

readStream.pipe(csv({headers: true, delimiter: options.delimiter}))
  .pipe(through2.obj(function(row, enc, cb){
    if (filterOptions.zipcodeFilter !== false && filterOptions.zipcodeFilter.indexOf(row[options.zipcodeCol]) === -1) {
      cb();
    } else if (filterOptions.stateFilter !== false && filterOptions.stateFilter.indexOf(row[options.stateCol]) === -1) {
      cb();
    } else {
      cb(null, row);
    }
  }))
  .pipe(grouper(70))
  // Coerce our parsed command-line args to an `Options` structure in a really
  // dubious fashion. We actually include far more keys than should appear in
  // `Options`.
  .pipe(geocoder(geocodeOptions))
  .pipe(writeStream);
